### PR TITLE
implement CosineEmbeddingLoss as a native function and add reduce arg

### PR DESCRIPTION
**Pull request by li-roy**
Added reduce arg as per #264.

forward (new) [1.1905965859768912, 1.160144692985341, 1.1558120870031416]
backward (new) [1.9150976981036365, 1.9792822760064155, 1.8779143309220672]
double backward (new) [3.6898688060464337, 3.5784677929477766, 3.569505032035522]

forward (old) [3.2359962839400396, 3.275224728975445, 3.3409753759624436]
backward (old) [5.668679727939889, 5.722980880062096, 5.585088661056943]
double backward (old) N/A

### CODE REVIEW

**Comment by zou3519** [comment]
Will review soon.

File: torch/nn/functional.py [hunk 1]
```diff
@@ -1636,12 +1636,12 @@ def multilabel_soft_margin_loss(input, target, weight=None, size_average=True, r
     return binary_cross_entropy(input, target, weight, size_average, reduce)
 
 
-def cosine_embedding_loss(input1, input2, target, margin=0, size_average=True):
-    """cosine_embedding_loss(input1, input2, target, margin=0, size_average=True) -> Variable
+def cosine_embedding_loss(input1, input2, target, margin=0, size_average=True, reduce=True):
+    """cosine_embedding_loss(input1, input2, target, margin=0, size_average=True, reduce=True) -> Variable
```
Comment (zou3519) [inline torch/nn/functional.py]: s/Variable/Tensor/

File: test/test_nn.py [hunk 2]
```diff
@@ -3765,6 +3765,20 @@ def test_pairwise_distance(self):
         input2 = Variable(torch.randn(4, 4), requires_grad=True)
         self.assertTrue(gradcheck(lambda x, y: F.pairwise_distance(x, y), (input1, input2)))
 
+    def test_cosine_embedding_loss_no_reduce(self):
+        input1 = Variable(torch.randn(15, 10), requires_grad=True)
+        input2 = Variable(torch.randn(15, 10), requires_grad=True)
+        target = Variable(torch.randn(15).sign())
+        self.assertTrue(gradcheck(lambda x, y, z: F.cosine_embedding_loss(
```
Comment (zou3519) [inline test/test_nn.py]: Could you make this into a module test? It tests some more things in addition to the gradcheck. It would also be nice to have a reference function test.

File: test/common_nn.py [hunk 3]
```diff
@@ -415,6 +415,25 @@ def multimarginloss_reference(input, target, p=1, margin=1, weight=None, size_av
         return output / dim
 
 
+def cosineembeddingloss_reference(input1, input2, target, margin=0, size_average=True, reduce=True):
+    cos = input1.new(input1.size(0)).zero_()
```
Comment (zou3519) [inline test/common_nn.py]: Can't you just write this as 
```
torch.where(target > 0, 1 - cos(input1, input2), max(0, cos(input1, input2) - margin)
```
and define the cos function somewhere?

File: torch/nn/modules/loss.py [hunk 4]
```diff
@@ -805,13 +805,14 @@ class CosineEmbeddingLoss(Module):
     batch samples. By default, `size_average = True`.
```
Comment (zou3519) [inline torch/nn/modules/loss.py]: Update the docs please!

File: aten/src/ATen/native/Loss.cpp [hunk 5]
```diff
@@ -1,9 +1,33 @@
...
 
+#define EPSILON 1e-12
+
 
 namespace at { namespace native {
 
+Tensor cosine_embedding_loss(const Tensor& input1, const Tensor& input2, const Tensor& target, double margin, bool size_average, bool reduce) {
+  auto prod_sum = (input1 * input2).sum(1);
+  auto mag_square1 = (input1 * input1).sum(1) + EPSILON;
+  auto mag_square2 = (input2 * input2).sum(1) + EPSILON;
```
Comment (zou3519) [inline aten/src/ATen/native/Loss.cpp]: Why is the epsilon needed?

File: aten/src/ATen/native/Loss.cpp [hunk 6]
```diff
@@ -1,9 +1,33 @@
...
+  auto denom = (mag_square1 * mag_square2).sqrt_();
+  auto cos = prod_sum / denom;
+
+  auto zeros = at::zeros_like(target);
+  auto pos = 1 - cos;
+  auto neg = (cos - margin).clamp_min_(0);
+  auto output_pos = at::where(target == 1, pos, zeros);
+  auto output_neg = at::where(target == -1, neg, zeros);
```
Comment (zou3519) [inline aten/src/ATen/native/Loss.cpp]: Should we throw an error if target is not -1 or 1? I'm asking because rewriting this as at::where(target > 0, pos, neg) is a little nicer.

**Review by zou3519 — COMMENTED** [review]
Math looks good. I had some minor comments. Please rebase and update the docs (can cosine_embedding_loss take in tensors of more than two dimensions? I think the answer is no but I'm not sure).

**Review by zou3519 — APPROVED** [review]
Looks good to me, let's wait for the CI to pass
